ci/docker_tool

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+
+IMAGE_NAME=lsinteg
+
+if [[ -z "$PROJECT" ]]; then 
+  PROJECT="ls_integ"
+fi
+
+echo "Using docker-compose project '$PROJECT'"
+
+RUN_PREFIX=docker
+LOGSTASH_VERSION=$(cat versions.yml  | egrep '^logstash:' | cut -d' ' -f 2)
+
+function startdev {
+  # If container exists, start it
+  docker ps -a | awk '{print $NF}' | grep lsdev
+  if [[ $? -eq 0 ]]; then
+    docker start lsdev
+  else # Otherwise run it
+    #-agentlib:jdwp=transport=dt_socket,server=y,address=8000,suspend=n
+    docker run \
+      -v "$PWD:/mnt/host:delegated" \
+      -e 'LS_JAVA_OPTS=-Xmx1500M -Xms1500M' \
+      -e 'INTEGRATION=true' \
+      -e "JRUBY_OPTS=-Xcompile.invokedynamic=false" \
+      -d \
+      --name lsdev \
+      $IMAGE_NAME \
+      bash -c ' tail -f /dev/null'
+  fi
+  # Speeds up gem install etc.
+  docker exec lsdev rsync /mnt/host/vendor /opt/logstash/vendor
+}
+
+function rsyncdev {
+  docker exec lsdev rsync --delete --exclude .git --exclude logs --exclude build --exclude .gradle --exclude vendor -r /mnt/host/ /opt/logstash/ 
+}
+
+function build {
+  docker build -t $IMAGE_NAME .
+}
+
+INTEGRATION_COMMAND="time $RUN_PREFIX run -t --rm $IMAGE_NAME ci/integration_run.sh"
+
+case "$1" in
+  "build") 
+    docker stop lsdev && \
+    docker rm lsdev && \
+    build
+  ;;
+  "ci_integration")
+    echo "Rebuilding images just in case"
+    build && \
+    $INTEGRATION_COMMAND ${@:2}
+  ;;
+  "ci_integration_parallel")
+    if [[ "$4" != "false" ]]; then # Build skip flag
+      build
+    fi
+    pushd qa/integration
+    spec_find_cmd='find specs -name *_spec.rb'
+    total_specs=$(expr $($spec_find_cmd | wc -l))
+    parallelism=$2
+    parallelism_index=$3
+
+    if [[ -z "$parallelism" || -z "$parallelism_index" ]]; then
+      echo "You must invoke with parallelism and index parameters. ex: ci/docker_tool ci_parallel_integration 4 1"
+      exit 1
+    fi
+
+    run_num_specs=$(expr $total_specs / $(expr $parallelism))
+    start_point=$(expr $run_num_specs \* $(expr $parallelism_index))
+    tail_back=$(expr $total_specs - $start_point)
+
+    if expr $(expr $parallelism_index + 1) = $parallelism; then
+      tail_back=$(expr $run_num_specs + $(expr $total_specs % $(expr $parallelism)))
+      run_num_specs=$tail_back
+    fi
+
+    echo "Running $run_num_specs specs for index "$parallelism_index" in $parallelism containers simultaneously"
+    echo "Total Specs: $total_specs"
+
+    matched_specs=$($spec_find_cmd | sort | tail -n $tail_back | head -n $run_num_specs | xargs -n 999999 echo)
+    popd
+    $INTEGRATION_COMMAND $matched_specs
+  ;;
+  "ci_integration_2x")
+    build
+    ci/docker_tool ci_integration_parallel 2 0 false &
+    ci/docker_tool ci_integration_parallel 2 1 false &
+
+    FAILED=0
+    for job in `jobs -p`
+    do
+    echo $job
+      wait $job || let "FAIL+=1"
+    done
+
+    if [[ "$FAILED" != "0" ]]; then
+      exit 1
+    fi
+    echo "Both jobs succeeded!"
```

Comment:
Hmm... this doesn't seem to work as expected [1].  Also, since the integration_run command is actually 2 commands [2] , I _think_ the logic here would only account for failures of the second command.  (I think that is is easily fixed though by checking the status of the first command and fail fast in integration_run.sh)

[1]
```
16:38:40 Finished in 2 minutes 23.9 seconds (files took 4.52 seconds to load)
16:38:40 3 examples, 1 failure
16:38:40 
16:38:40 Failed examples:
16:38:40 
16:38:40 rspec ./specs/cli/prepare_offline_pack_spec.rb:46 # CLI > logstash-plugin prepare-offline-pack create a pack from a wildcard successfully create a pack
16:38:40 
16:38:40 Randomized with seed 57507
16:38:40 
16:38:40 0.04user 0.09system 23:48.03elapsed 0%CPU (0avgtext+0avgdata 11124maxresident)k
16:38:40 188536inputs+0outputs (297major+5862minor)pagefaults 0swaps
16:38:40 32392
16:38:40 Both jobs succeeded!
```

[2]
```
bundle exec rspec -fd --tag ~offline $@
bundle exec rspec -fd --tag offline $@
```